### ISSUE TITLE
git_clone() with SSH URL can randomly fail with 'would block' errors

### ISSUE THREAD

**Issue by mgorny** [libgit2/libgit2#3563]
It looks that SSH transport reads/writes are not accouting for possible `LIBSSH2_ERROR_EAGAIN`, which means that they can fail when the syscall is interrupted by a signal. This is quite a nuisance when cloning large repositories.

I've attempted the 'obvious' fix of looping when `LIBSSH2_ERROR_EAGAIN` is received. Sadly, it didn't fix the issue in my program and instead caused it to hang randomly when fetching. So I'm reporting a bug instead, and hoping someone could come up with something better. I'll try to set up a proper test case later.

**Comment by carlosmn** [comment]
The libssh2 documentation does not say `_EAGAIN` is returned when interrupted by a signal, but when the call would block. We set the channels to blocking mode so we don't have to bother with this kind of issue. We do want to block, and libssh2 should not be returning this at all.

This looks like a duplicate of #3382

**Comment by mgorny** [comment]
That would depend on how you define 'blocking'. All POSIX I/O facilities (i.e. `read()`, `write()`, `close()`) can be interrupted by signals when in blocking mode, and return EAGAIN in this case. Having libssh2 behave differently would be at least unexpected.

**Comment by ethomson** [comment]
That's a very theoretical claim.  In practice:  libssh2 is returning `EAGAIN` and it hangs when you try again.  Not really sure what we could possibly do here?  This sounds very much like #3382.

**Comment by carlosmn** [comment]
The claim in the libssh2 documentation is that `_EAGAIN` is for non-blocking behaviour. If they're returning it in other cases either the documentation or the implementation are wrong. Unless they specify what they expect the user to handle, whatever we do is going to be a guess.

Looking though the libssh2 code, their transport code does seem to convert `EINTR` to `_EAGAIN`, but the ssh code proper (which delegates reading and writing to the transport code) is assuming that libssh2's `_EAGAIN` means the system's `EAGAIN`/`EWOULDBLOCK` and sets the error message saying it would block, which gives us the non-sensical error message saying it would block. When what's happening is that we need to call the function again with the same arguments.

This is rather unfortunate, since there's not mention of this behaviour in the documentation, and even libssh2 code does not seem aware of this as the error message attests. We should bring this up with the libssh2 project and wrap our calls with a `do { error = X() } while (error == LIBSSH2_EAGAIN);`. But if that blocks, then I'm not sure what to do. The examples in the libssh2 codebase don't really seem to make use of the blocking mode, so maybe we should try setting them as non-blocking and wait on `select(2)` like we do for curl.

**Comment by magcius** [comment]
@mgorny read / write syscalls return EINTR when interrupted by signals, not EAGAIN.

**Comment by mgorny** [comment]
Yes, sorry, my thinko.
